algorithms/nautilus/fuzzer/fuzzer.cpp

```diff
@@ -0,0 +1,278 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+/**
+ * @file fuzzer.cpp
+ * @brief Fuzzing loop of Nautilus.
+ * @author Ricerca Security <fuzzuf-dev@ricsec.co.jp>
+ */
+#include <fstream>
+#include <memory>
+#include <nlohmann/json.hpp>
+#include <string>
+#include <vector>
+#include "fuzzuf/algorithms/nautilus/fuzzer/fuzzer.hpp"
+#include "fuzzuf/algorithms/nautilus/fuzzer/mutation_hierarflow_routines.hpp"
+#include "fuzzuf/algorithms/nautilus/fuzzer/other_hierarflow_routines.hpp"
+#include "fuzzuf/algorithms/nautilus/fuzzer/update_hierarflow_routines.hpp"
+#include "fuzzuf/hierarflow/hierarflow_intermediates.hpp"
+#include "fuzzuf/hierarflow/hierarflow_node.hpp"
+#include "fuzzuf/hierarflow/hierarflow_routine.hpp"
+#include "fuzzuf/logger/logger.hpp"
+#include "fuzzuf/utils/common.hpp"
+#include "fuzzuf/utils/filesystem.hpp"
+
+
+namespace fuzzuf::algorithm::nautilus::fuzzer {
+
+using json = nlohmann::json;
+
+/**
+ * @fn
+ * @brief Construct Nautilus fuzzer
+ * @param (state_ref) Reference to the state of Nautilus fuzzer
+ */
+NautilusFuzzer::NautilusFuzzer(std::unique_ptr<NautilusState>&& state_ref)
+  : state(std::move(state_ref)) {
+  /* Check files and load grammar */
+  CheckPathExistence();
+  LoadGrammar();
+
+  /* Construct fuzzing loop */
+  BuildFuzzFlow();
+
+  /* Clear screen */
+  MSG(TERM_CLEAR);
+}
+
+/**
+ * @fn
+ * @brief Build HierarFlow of Nautilus
+ */
+void NautilusFuzzer::BuildFuzzFlow() {
+  using fuzzuf::hierarflow::CreateNode;
+  using namespace fuzzuf::algorithm::nautilus::fuzzer::routine::other;
+  using namespace fuzzuf::algorithm::nautilus::fuzzer::routine::mutation;
+  using namespace fuzzuf::algorithm::nautilus::fuzzer::routine::update;
+
+  fuzz_loop = CreateNode<FuzzLoop>(*state);
+
+  /* Main flow */
+  auto select_input     = CreateNode<SelectInput>(*state);
+  auto process_input_or = CreateNode<ProcessInput>(*state);
+  auto generate_input   = CreateNode<GenerateInput>(*state);
+  auto update_state     = CreateNode<UpdateState>(*state);
+
+  /* Processing flow */
+  auto initialize_state_or = CreateNode<InitializeState>(*state);
+  auto apply_det_muts_or   = CreateNode<ApplyDetMuts>(*state);
+  auto apply_rand_muts     = CreateNode<ApplyRandMuts>(*state);
+
+  /* Mutation flow */
+  auto splice    = CreateNode<MutSplice>(*state);
+  auto havoc     = CreateNode<MutHavoc>(*state);
+  auto havoc_rec = CreateNode<MutHavocRec>(*state);
+
+  fuzz_loop << (
+    select_input << (
+      process_input_or
+      || generate_input // TODO: maybe execute.HardLink() here
+    )
+    || update_state
+  );
+
+  process_input_or << (
+    initialize_state_or
+    || apply_det_muts_or << (
+      splice.HardLink()
+      || havoc.HardLink()
+      || havoc_rec.HardLink()
```

Comment:
Should this really be included in det muts? 
I saw the same in the original implementation, but its paper says different. I'm puzzled.

Reply:
According to the section V of the original paper, only the difference between deterministic phase and random phase is whether it runs the Rules Mutation. The Rules Mutation is called by `state.DeterministicTreeMutation` in `ApplyDetMuts`, while it's not called in `ApplyRandMuts`.
I think both this and the original implementation is correct.
Be noted that detafl is not implemented in the original Nautilus too.

Reply:
Oh I overlooked `deterministic_tree_mutation` for some reason....
Maybe, the difference betweeen det muts and rand muts would be clearer if you make `ApplyDetMuts` only call successers just like `ApplyRandMuts` and move the call of `DeterministicTreeMutation` into one of its successor.

Reply:
Modified in [cec3057](https://github.com/fuzzuf/fuzzuf/pull/23/commits/cec3057c39133b10e273fd75344f4955843e70d4)